File: go/metrics.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// MetricsListMetricsOptions carries the query and header parameters of
// Metrics.ListMetrics.
//
// Optional parameters are pointers; leave them nil to omit them.
type MetricsListMetricsOptions struct {
	ProductFamilyId *ProductFamilyId

	// Search by metric name or code
	Search *string

	// Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`.
	OrderBy *string

	// Page number (0-indexed)
	Page *int32

	// Number of items per page
	PerPage *int32
}

// Metrics groups the metrics operations of the Meteroid API.
type Metrics struct {
	client *Client
}

func (a *Metrics) ListMetrics(ctx context.Context, options *MetricsListMetricsOptions) (*MetricListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/metrics")

	if options != nil {
		if options.ProductFamilyId != nil {
			req.SetQueryParam("product_family_id", string(*options.ProductFamilyId))
		}
		if options.Search != nil {
			req.SetQueryParam("search", *options.Search)
		}
		if options.OrderBy != nil {
			req.SetQueryParam("order_by", *options.OrderBy)
		}
		if options.Page != nil {
			req.SetQueryParam("page", formatInt(int64(*options.Page)))
		}
		if options.PerPage != nil {
			req.SetQueryParam("per_page", formatInt(int64(*options.PerPage)))
		}
	}

	var out MetricListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Metrics) CreateMetric(ctx context.Context, createMetricRequest CreateMetricRequest) (*Metric, error) {
	req := newRequest(http.MethodPost, "/api/v1/metrics")

	req.SetJSONBody(createMetricRequest)

	var out Metric
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Metrics) GetMetric(ctx context.Context, metricId string) (*Metric, error) {
	req := newRequest(http.MethodGet, "/api/v1/metrics/{metric_id}")
	req.SetPathParam("metric_id", metricId)

	var out Metric
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Partially update metric fields. Code and aggregation_type are immutable.
func (a *Metrics) UpdateMetric(ctx context.Context, metricId string, updateMetricRequest UpdateMetricRequest) (*Metric, error) {
	req := newRequest(http.MethodPatch, "/api/v1/metrics/{metric_id}")
	req.SetPathParam("metric_id", metricId)

	req.SetJSONBody(updateMetricRequest)

	var out Metric
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Metrics) ArchiveMetric(ctx context.Context, metricId string) error {
	req := newRequest(http.MethodPost, "/api/v1/metrics/{metric_id}/archive")
	req.SetPathParam("metric_id", metricId)

	return a.client.execute(ctx, req, nil)
}

func (a *Metrics) UnarchiveMetric(ctx context.Context, metricId string) error {
	req := newRequest(http.MethodPost, "/api/v1/metrics/{metric_id}/unarchive")
	req.SetPathParam("metric_id", metricId)

	return a.client.execute(ctx, req, nil)
}

File: python/meteroid/api/metrics.py
# this file is @generated
import typing as t

from ..models import (
    CreateMetricRequest,
    Metric,
    MetricListResponse,
    ProductFamilyId,
    UpdateMetricRequest,
)
from .common import ApiBaseAsync, ApiBaseSync, decode_response, serialize_query_params


class MetricsAsync(ApiBaseAsync):
    """metrics API."""

    async def list_metrics(
        self,
        *,
        product_family_id: t.Optional[ProductFamilyId] = None,
        search: t.Optional[str] = None,
        order_by: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
    ) -> MetricListResponse:
        """:param search: Search by metric name or code
        :param order_by: Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/metrics",
            query_params=serialize_query_params(
                {
                    "product_family_id": product_family_id,
                    "search": search,
                    "order_by": order_by,
                    "page": page,
                    "per_page": per_page,
                },
            ),
        )
        return decode_response(response, MetricListResponse)

    async def create_metric(
        self,
        create_metric_request: CreateMetricRequest,
    ) -> Metric:
        response = await self._request_asyncio(
            method="post",
            path="/api/v1/metrics",
            json_body=create_metric_request.to_dict(),
        )
        return decode_response(response, Metric)

    async def get_metric(
        self,
        metric_id: str,
    ) -> Metric:
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/metrics/{metric_id}",
            path_params={
                "metric_id": metric_id,
            },
        )
        return decode_response(response, Metric)

    async def update_metric(
        self,
        metric_id: str,
        update_metric_request: UpdateMetricRequest,
    ) -> Metric:
        """Partially update metric fields. Code and aggregation_type are immutable."""
        response = await self._request_asyncio(
            method="patch",
            path="/api/v1/metrics/{metric_id}",
            path_params={
                "metric_id": metric_id,
            },
            json_body=update_metric_request.to_dict(),
        )
        return decode_response(response, Metric)

    async def archive_metric(
        self,
        metric_id: str,
    ) -> None:
        await self._request_asyncio(
            method="post",
            path="/api/v1/metrics/{metric_id}/archive",
            path_params={
                "metric_id": metric_id,
            },
        )

    async def unarchive_metric(
        self,
        metric_id: str,
    ) -> None:
        await self._request_asyncio(
            method="post",
            path="/api/v1/metrics/{metric_id}/unarchive",
            path_params={
                "metric_id": metric_id,
            },
        )


class Metrics(ApiBaseSync):
    """metrics API."""

    def list_metrics(
        self,
        *,
        product_family_id: t.Optional[ProductFamilyId] = None,
        search: t.Optional[str] = None,
        order_by: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
    ) -> MetricListResponse:
        """:param search: Search by metric name or code
        :param order_by: Sort order. Format: `column.direction`. Allowed columns: `name`, `code`, `created_at`. Direction: `asc` or `desc`. Default: `name.asc`.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = self._request_sync(
            method="get",
            path="/api/v1/metrics",
            query_params=serialize_query_params(
                {
                    "product_family_id": product_family_id,
                    "search": search,
                    "order_by": order_by,
                    "page": page,
                    "per_page": per_page,
                },
            ),
        )
        return decode_response(response, MetricListResponse)

    def create_metric(
        self,
        create_metric_request: CreateMetricRequest,
    ) -> Metric:
        response = self._request_sync(
            method="post",
            path="/api/v1/metrics",
            json_body=create_metric_request.to_dict(),
        )
        return decode_response(response, Metric)

    def get_metric(
        self,
        metric_id: str,
    ) -> Metric:
        response = self._request_sync(
            method="get",
            path="/api/v1/metrics/{metric_id}",
            path_params={
                "metric_id": metric_id,
            },
        )
        return decode_response(response, Metric)

    def update_metric(
        self,
        metric_id: str,
        update_metric_request: UpdateMetricRequest,
    ) -> Metric:
        """Partially update metric fields. Code and aggregation_type are immutable."""
        response = self._request_sync(
            method="patch",
            path="/api/v1/metrics/{metric_id}",
            path_params={
                "metric_id": metric_id,
            },
            json_body=update_metric_request.to_dict(),
        )
        return decode_response(response, Metric)

    def archive_metric(
        self,
        metric_id: str,
    ) -> None:
        self._request_sync(
            method="post",
            path="/api/v1/metrics/{metric_id}/archive",
            path_params={
                "metric_id": metric_id,
            },
        )

    def unarchive_metric(
        self,
        metric_id: str,
    ) -> None:
        self._request_sync(
            method="post",
            path="/api/v1/metrics/{metric_id}/unarchive",
            path_params={
                "metric_id": metric_id,
            },
        )
